### PR TITLE
Negotiate downttrack for subscriber before receiver is ready

### PR DESCRIPTION
**Pull request by cnderrauber**
This change will save 1 round sdp negotiation time for subscribing to simulcast-codec or remote node track

### CODE REVIEW

File: pkg/sfu/receiver.go [hunk 1]
```diff
@@ -841,6 +844,11 @@ func (w *WebRTCReceiver) GetMonotonicNowUnixNano() int64 {
 	return w.baseTime.Add(time.Since(w.baseTime)).UnixNano()
 }
 
+func (w *WebRTCReceiver) AddOnReady(fn func()) {
+	// webRTCReceiver is always ready after created
+	fn()
```
Comment (boks1971) [inline pkg/sfu/receiver.go:849]: Are there any locking issues here? What if caller is holding some lock and the callback also needs that lock?

File: pkg/sfu/downtrack.go [hunk 2]
```diff
@@ -1935,7 +2012,7 @@ func (d *DownTrack) DebugInfo() map[string]interface{} {
 		"StreamID":            d.params.StreamID,
 		"SSRC":                d.ssrc,
 		"MimeType":            d.codec.MimeType,
-		"Bound":               d.bound.Load(),
+		"Bound":               d.bindState.Load(),
```
Comment (boks1971) [inline pkg/sfu/downtrack.go:2015]: Can just change this to `BindState` and make a `String()` method for `bindState` to make it more readable?

File: pkg/sfu/downtrack.go [hunk 3]
```diff
@@ -481,18 +448,129 @@ func (d *DownTrack) Bind(t webrtc.TrackLocalContext) (webrtc.RTPCodecParameters,
...
+	}
+}
+
 // Unbind implements the teardown logic when the track is no longer needed. This happens
 // because a track has been stopped.
 func (d *DownTrack) Unbind(_ webrtc.TrackLocalContext) error {
 	d.bindLock.Lock()
-	d.bound.Store(false)
-	d.onBindAndConnectedChange()
+	d.bindState.Store(bindStateUnbound)
```
Comment (boks1971) [inline pkg/sfu/downtrack.go:573]: guess this should use `setBindStateLocked` so that `onBindAndConnectedChange` is invoked?

**Review by boks1971 — APPROVED** [review]
nice!!!

File: pkg/sfu/receiver.go (hunk 1 above)
Comment (cnderrauber) [reply pkg/sfu/receiver.go:849]: It is ok that the downtrack doesn't hold lock when calling `AddOnReady`, will add a comment for future that the callback will be called immediately if receiver is already ready.

File: pkg/sfu/downtrack.go (hunk 3 above)
Comment (cnderrauber) [reply pkg/sfu/downtrack.go:573]: good catch!

**Pull request merged**